dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,232 @@
+    }
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  while(ReadFlushFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = false;
+    entry.pts = frame_->pts;
+    entry.is_flush_frame = true;
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  Reset();
+}
+
+void VideoFileCPU::CopyToOutput(uint8_t *data) {
+  dest_[0] = data;
+  dest_linesize_[0] = frame_->width * Channels();
+  if (sws_scale(sws_ctx_, frame_->data, frame_->linesize, 0, frame_->height, dest_, dest_linesize_) < 0) {
+    DALI_FAIL("Could not convert frame data to RGB");
+  }
+}
+
+bool VideoFileCPU::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  while(av_read_frame(ctx_, packet_) >= 0) {
+    if (packet_->stream_index != stream_id_) {
+      continue;
+    }
+
+    if (avcodec_send_packet(codec_ctx_, packet_) < 0) {
+      DALI_FAIL("Failed to send packet to decoder");
+    }
+
+    int ret = avcodec_receive_frame(codec_ctx_, frame_);
+
+    if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+      continue;
+    }
+
+    if (!copy_to_output) {
+      return true;
+    }
+
+    if (sws_ctx_ == nullptr) {
+      sws_ctx_ = sws_getContext(
+        Width(), 
+        Height(), 
+        codec_ctx_->pix_fmt, 
+        Width(), 
+        Height(), 
+        AV_PIX_FMT_RGB24, 
+        SWS_BILINEAR, 
+        nullptr, 
+        nullptr, 
+        nullptr);
+      DALI_ENFORCE(sws_ctx_, "Could not create sw context");
+    }
+
+    CopyToOutput(data);
+    return true;
+  }
+
+  return false;
+}
+
+void VideoFileCPU::Reset() {
+  DALI_ENFORCE(
+    av_seek_frame(ctx_, stream_id_, 0, AVSEEK_FLAG_FRAME) >= 0,
+    make_string("Could not seek to the first frame of video ", filename_));
+  avcodec_flush_buffers(codec_ctx_);
+}
+
+void VideoFileCPU::SeekFrame(int frame_id) {
+  // TODO(awolant): Optimize seeking: 
+  //  - when we seek next frame,
+  //  - when we seek frame with the same keyframe as the current frame
+  //  - for CFR, when we know pts, but don't know keyframes 
+
+  DALI_ENFORCE(
+    frame_id >= 0 && frame_id < NumFrames(),
+    make_string("Invalid seek frame id. frame_id = ", frame_id, ", num_frames = ", NumFrames()));
+
+  auto &frame_entry = index_[frame_id];
+  int keyframe_id = frame_entry.last_keyframe_id;
+  auto &keyframe_entry = index_[keyframe_id];
+
+  // Seeking while on flush frame. Need to reset flush state
+  if (flush_state_) {
+    flush_state_ = false;
+  }
+ 
+  DALI_ENFORCE(
+    av_seek_frame(ctx_, stream_id_, keyframe_entry.pts, AVSEEK_FLAG_FRAME) >= 0,
+    make_string("Failed to seek to frame ", frame_id, "with keyframe", keyframe_id, "in video ", filename_));
+
+  avcodec_flush_buffers(codec_ctx_);
+
+  for (int i = 0; i < (frame_id - keyframe_id); ++i) {
+    ReadNextFrame(nullptr, false);
+  }
+}
+
+bool VideoFileCPU::ReadFlushFrame(uint8_t *data, bool copy_to_output) {
+  if (!flush_state_) {
+    avcodec_send_packet(codec_ctx_, nullptr);
```

Comment:
So ReadRegularFrame leaves one packet that was not sent to the decoder? I would rather assume that `ReadRegularFrame` should set flush_state_ when it runs out of frames.

Reply:
Done, moved entering the flush state to the `ReadRegularFrame`